add new supply-chain article (#1127)
returntocorp/semgrep-docs: docs/kb/semgrep-supply-chain/scanning-projects-with-multiple-lock-files.md

---
tags:
  - GitHub Actions
  - Lock files
description: Scanning projects with multiple lock files in GitHub Actions
---

# Scanning projects with multiple lock files in GitHub Actions

> Your project may have many modules, each with a dependency descriptor file; for example, `pom.xml` if Maven is used as a build dependency tool.

Reviewer: For "dependency descriptor file" do you mean a manifest file or a lock file?

Author: For example, for Maven I would say:
- pom.xml is a dependency descriptor file or manifest file
- maven_dep_tree.txt is the lock file

Reviewer: Ok, I'll just suggest to prefer "manifest file" because it's more common :)

> This [project](https://github.com/finos/legend-engine) complies with the above description.

Reviewer 2: What's the purpose of this link? If it's just to show what such a project might look like, think we probably shouldn't link out to random GH projects in our KB/docs for that. If we need an example, creating one we control would be better.

Author: You are right. I have already changed the link.

> In this case, Semgrep will run a successful supply chain scan of all the lock files, one per module, if they are correctly generated.

Reviewer 2: ```suggestion
If you correctly generate all lock files, one per module, Semgrep can perform a successful Supply Chain scan for all dependencies in the project.
```

> ## Generating lock files as the previous step

Reviewer 2: ```suggestion
## Generating lock files before scanning
```

In the case of using Maven, the command to use is:
```
mvn dependency:tree -DoutputFile=maven_dep_tree.txt
```

> And this step must be executed before calling Semgrep. So, a typical Jenkins pipeline can look like this:

Reviewer 2: ```suggestion
This step must be executed before scanning with Semgrep. Here is an example of a Jenkins pipeline executing this step:
```

```
pipeline {
  agent any
    environment {
      SEMGREP_APP_TOKEN = credentials('SEMGREP_APP_TOKEN')
      SEMGREP_BRANCH="${GIT_BRANCH}"
    }
    stages {
      stage ('Generate-LockFile') {
        steps {
            withMaven(maven: 'maven') {
              sh "mvn dependency:tree -DoutputFile=maven_dep_tree.txt"
            }
        }
      }
      stage('Semgrep-Scan') {
        steps {
                script {
                    if (env.GIT_BRANCH == 'master') {
                      sh 'pip3 install semgrep'
                      sh 'semgrep ci'
                    }  
                }
        }
      }
    }
}
```

> In the case of using GitHub Actions as a CI platform, then using artifacts to share files between jobs is required.

Reviewer 2: ```suggestion
If you are using GitHub Actions as a CI platform, using artifacts to share files between jobs is required.
```
I'd also move this line down to the next section where you talk about the process - otherwise it's a bit odd to give a Jenkins example and then talk about GitHub Actions.

## Sharing lock files as artifacts
As the project can have multiple dependency descriptor files (`pom.xml` in the case of Maven), there will be multiple lock files (`maven_dep_tree.txt`).

> These lock files must be shared as artifacts between jobs, and the efficient way to do it are through a zip file that can gather all lock files and then, in the next job, unzip the lock files and run Semgrep as usual.

Reviewer 2: ```suggestion
These lock files must be shared as artifacts between jobs, and the efficient way to do it is through a zip file that gathers all lock files and, in the next job, unzips the lock files and runs Semgrep as usual.
```

> A GitHub Actions pipeline can look like this:

Reviewer 2: ```suggestion
Here is a GitHub Actions workflow example:
```

  schedule:
  - cron: 00 5 1 * *
name: Semgrep
jobs:
  buildmavenDepTree: 
      runs-on: ubuntu-latest
      steps:
      - uses: actions/checkout@v3
      - name: Set up JDK 11
        uses: actions/setup-java@v3
        with:
          java-version: '11'
          distribution: 'temurin'
      - name: Build Dependency Tree
        run: mvn dependency:tree -DoutputFile=maven_dep_tree.txt -Dmaven.test.skip=true
      - name: Create Zip File
        run: find . -type f -name 'maven_dep_tree.txt' -exec zip -r archive.zip {} +
      - name: Upload Dependency Zip    
        uses: actions/upload-artifact@v3
        with:
          name: zipfile
          path: archive.zip
  semgrep:
    needs: buildmavenDepTree
    name: Scan
    runs-on: ubuntu-20.04
    env:
      SEMGREP_APP_TOKEN: ${{ secrets.SEMGREP_APP_TOKEN }}
    container:
      image: returntocorp/semgrep
    steps:
      - uses: actions/checkout@v3
      - name: Download artifact from the previous job
        uses: actions/download-artifact@v3
        with:
          name: zipfile    
      - name: Semgrep Scan  
        run: |
          unzip -o archive.zip
          semgrep ci
```

## Conclusions

> In the case of using GitHub Actions and if your project has multiple lock files, uploading all of them as a zip file and downloading them in the next step can help to get a successful semgrep supply chain scan.

Reviewer 3: would it be helpful to specify here again that this is only for java/maven?

Reviewer 2: I think probably this whole article should specify that, unless I'm misunderstanding the scope. So the title would be "Scanning Apache Maven projects with multiple lockfiles" or something of that nature.

Author: You are right. I have changed the title and conclusion to specify it is for GHA and Maven. But, I would add, in conclusion, that it can help in other situations:

_This concept (shared lock files as zip files between jobs) can be extrapolated to other build dependencies systems such as Gradle or other platforms such as GitLab, Bitbucket, Jenkins, etc..._

---
tags:
  - GitHub Actions
  - Lock files
description: Scanning projects with multiple lock files: a Maven example
---

# Scanning Apache Maven projects with multiple lockfiles

> Your project may have many modules, each with a dependency descriptor file; for example, `pom.xml` if Maven is used as a build dependency tool.

Reviewer: ```suggestion
Your project may have many modules, each with a manifest file; for example, `pom.xml` if Maven is used as a build dependency tool.
```
